tests/unit/request/xhr.ts

```diff
@@ -2,168 +2,497 @@ import * as registerSuite from 'intern!object';
+	}
+}
+
+let echoServerAvailable = false;
 registerSuite({
 	name: 'request/xhr',
 
+	before: function () {
+		return new Promise(function (resolve, reject) {
+			xhrRequest('/__echo/', {
+				method: 'get',
+				timeout: 10000
+			}).then(
+				function (response) {
+					if (response && response.statusCode === 200) {
+						echoServerAvailable = true;
+					}
+					resolve();
+				},
+				function () {
+					resolve();
+				}
+			);
+		});
+	},
+
 	'HTTP methods': {
-		specified(): void {
-			const dfd = this.async();
-			xhrRequest(getRequestUrl('foo.json'), { method: 'get' })
+		specified() {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest(getRequestUrl('foo.json'), { method: 'get' })
 				.then(
-					dfd.callback(function (response: any): void {
+					function (response: any) {
 						assert.strictEqual(response.requestOptions.method, 'get');
-					}),
-					dfd.reject.bind(dfd)
+					}
 				);
 		},
 
-		'default'(): void {
-			const dfd = this.async();
-			xhrRequest(getRequestUrl('foo.json'))
+		'default'() {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest(getRequestUrl('foo.json'))
 				.then(
-					dfd.callback(function (response: any): void {
+					function (response: any) {
 						assert.strictEqual(response.requestOptions.method, 'GET');
-					}),
-					dfd.reject.bind(dfd)
+					}
 				);
+		},
+
+		'.get with URL query'() {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest('/__echo/xhr?color=blue&numbers=one&numbers=two', {
+				handleAs: 'json'
+			}).then(
+				function (response: any) {
+					const query = JSON.parse(response.data).query;
+					assert.strictEqual(query.color, 'blue');
+					assert.deepEqual(query.numbers, [ 'one', 'two' ]);
+					assert.strictEqual(
+						response.url,
+						'/__echo/xhr?color=blue&numbers=one&numbers=two'
+					);
+				}
+			);
+		},
+
+		'.post': function () {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest('/__echo/post', {
+				method: 'POST',
+				data: new UrlSearchParams({ color: 'blue' }).toString(),
+				responseType: 'json'
+			}).then(
+				function (response: any) {
+					const data = JSON.parse(response.data);
+					assert.strictEqual(data.method, 'POST');
+					const payload = data.payload;
+
+					assert.ok(payload && payload.color);
+					assert.strictEqual(payload.color, 'blue');
+				}
+			);
 		}
 	},
 
 	'request options': {
-		'timeout'(): void {
-			const dfd = this.async();
-			xhrRequest(getRequestUrl('foo.json'), { timeout: 1 })
+		'timeout'() {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest(getRequestUrl('foo.json'), { timeout: 1 })
 				.then(
-					dfd.resolve.bind(dfd),
-					dfd.callback(function (error: Error): void {
+					null,
```

Comment:
I added this, and the test immediately started failing. It was resolving successfully, but with a response of 0 whenever it timed out. Despite being an invalid HTTP status, 0 is a valid value for statusCode in an XMLHttpRequest, and indicates an error of some sort, but I'm not sure why it was consistently firing when the request timed out. 

Wrapping the resolution of the request in src/request/xhr.ts with a check to make sure the status code was between 200 and 400 fixes it, but presumably this will make the request hang if the status code is not in that range.